src/libs/Middleware/Pagination.ts

```diff
@@ -0,0 +1,127 @@
+// TODO: Is this a legit use case for exposing `OnyxCache`, or should we use `Onyx.connect`?
+import fastMerge from 'expensify-common/dist/fastMerge';
+import type {OnyxCollection} from 'react-native-onyx';
+import Onyx from 'react-native-onyx';
+import type {ApiCommand} from '@libs/API/types';
+import Log from '@libs/Log';
+import PaginationUtils from '@libs/PaginationUtils';
+import CONST from '@src/CONST';
+import type {OnyxCollectionKey, OnyxPagesKey, OnyxValues} from '@src/ONYXKEYS';
+import type {Request} from '@src/types/onyx';
+import type {PaginatedRequest} from '@src/types/onyx/Request';
+import type Middleware from './types';
+
+type PaginationConfig<TResourceKey extends OnyxCollectionKey, TPageKey extends OnyxPagesKey> = {
+    initialCommand: ApiCommand;
+    previousCommand: ApiCommand;
+    nextCommand: ApiCommand;
+    resourceCollectionKey: TResourceKey;
+    pageCollectionKey: TPageKey;
+    sortItems: (items: OnyxValues[TResourceKey]) => Array<OnyxValues[TResourceKey] extends Record<string, infer TResource> ? TResource : never>;
+    getItemID: (item: OnyxValues[TResourceKey] extends Record<string, infer TResource> ? TResource : never) => string;
+};
+
+type PaginationConfigMapValue = Omit<PaginationConfig<OnyxCollectionKey, OnyxPagesKey>, 'initialCommand' | 'previousCommand' | 'nextCommand'> & {
+    type: 'initial' | 'next' | 'previous';
+};
+
+const paginationConfigs = new Map<string, PaginationConfigMapValue>();
+const ressources = new Map<OnyxCollectionKey, OnyxCollection<OnyxValues[OnyxCollectionKey]>>();
+const pages = new Map<OnyxPagesKey, OnyxCollection<OnyxValues[OnyxPagesKey]>>();
+
+function registerPaginationConfig<TResourceKey extends OnyxCollectionKey, TPageKey extends OnyxPagesKey>({
```

Comment:
It will be easier to work with if you are having TS issues:
```suggestion
function registerPaginationConfig<TResourceKey extends OnyxCollectionKey, TPageKey extends OnyxPagesKey>({
    initialCommand,
    previousCommand,
    nextCommand,
    resourceCollectionKey,
    pageCollectionKey,
    sortItems,
    getItemID,
}: PaginationConfig<TResourceKey, TPageKey>): void {
```

Reply:
choosing to ignore this one since, where the `...config` is used we're doing an unsafe `as unknown as X` cast anyways.